source/windows/bcrypt_rsa.c

```diff
@@ -0,0 +1,458 @@
+    size_t key_size_in_bits) {
+
+    aws_thread_call_once(&s_rsa_thread_once, s_load_alg_handle, NULL);
+
+    if (key_size_in_bits < AWS_CAL_RSA_MIN_SUPPORTED_KEY_SIZE_IN_BITS ||
+        key_size_in_bits > AWS_CAL_RSA_MAX_SUPPORTED_KEY_SIZE_IN_BITS) {
+        aws_raise_error(AWS_ERROR_INVALID_STATE);
+        return NULL;
+    }
+
+    struct bcrypt_rsa_key_pair *key_impl = aws_mem_calloc(allocator, 1, sizeof(struct bcrypt_rsa_key_pair));
+
+    aws_ref_count_init(&key_impl->base.ref_count, &key_impl->base, aws_rsa_key_pair_destroy);
+    key_impl->base.impl = key_impl;
+    key_impl->base.allocator = allocator;
+
+    NTSTATUS status = BCryptGenerateKeyPair(s_rsa_alg, &key_impl->key_handle, (ULONG)key_size_in_bits, 0);
+
+    if (!BCRYPT_SUCCESS(status)) {
+        aws_raise_error(AWS_ERROR_CAL_CRYPTO_OPERATION_FAILED);
+        goto on_error;
+    }
+
+    status = BCryptFinalizeKeyPair(key_impl->key_handle, 0);``
+
+    if (!BCRYPT_SUCCESS(status)) {
+        aws_raise_error(AWS_ERROR_CAL_CRYPTO_OPERATION_FAILED);
+        goto on_error;
+    }
+
+    /*
+     * bcrypt only exports to blob format, which is missing a few of the rsa
+     * numbers needed to reconstruct pkcs1 format. lets not init key buffers for
+     * now. this means we cannot support retrieving underlying key on win.
+     * TODO: bcrypt workarounds to retrieve/compute missing things? old crypt lib
+     * seems to have functionality we need.
+     */
+    AWS_ZERO_STRUCT(key_impl->base.priv);
+    AWS_ZERO_STRUCT(key_impl->base.pub);
+
+    key_impl->base.vtable = &s_rsa_key_pair_vtable;
+    key_impl->base.key_size_in_bits = key_size_in_bits;
+
+    return &key_impl->base;
+
+on_error:
+    s_rsa_destroy_key(&key_impl->base);
+    return NULL;
+}
+
+struct aws_rsa_key_pair *aws_rsa_key_pair_new_from_private_key_pkcs1_impl(
+    struct aws_allocator *allocator,
+    struct aws_byte_cursor key) {
+
+    aws_thread_call_once(&s_rsa_thread_once, s_load_alg_handle, NULL);
+    struct bcrypt_rsa_key_pair *key_pair_impl = aws_mem_calloc(allocator, 1, sizeof(struct bcrypt_rsa_key_pair));
+
+    aws_ref_count_init(&key_pair_impl->base.ref_count, &key_pair_impl->base, aws_rsa_key_pair_destroy);
+    key_pair_impl->base.impl = key_pair_impl;
+    key_pair_impl->base.allocator = allocator;
+    aws_byte_buf_init_copy_from_cursor(&key_pair_impl->base.priv, allocator, key);
+
+    struct aws_der_decoder *decoder = aws_der_decoder_new(allocator, key);
+
+    if (!decoder) {
+        goto on_error;
+    }
+
+    struct aws_rsa_private_key_pkcs1 private_key_data;
+    AWS_ZERO_STRUCT(private_key_data);
+    if (aws_der_decoder_load_private_rsa_pkcs1(decoder, &private_key_data)) {
+        goto on_error;
+    }
+
+    /* Hard to predict final blob size, so use pkcs1 key size as upper bound. */
+    size_t total_buffer_size = key.len + sizeof(BCRYPT_RSAKEY_BLOB);
+
+    if (aws_byte_buf_init(&key_pair_impl->key_buf, allocator, total_buffer_size)) {
+        goto on_error;
+    }
```

Comment:
trivial: allocation can't fail
```suggestion
    aws_byte_buf_init(&key_pair_impl->key_buf, allocator, total_buffer_size));
```